File: order-processor.js
const express = require('express');
const cors = require('cors');
const axios = require('axios');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

const app = express();
const PORT = process.env.PORT || 10001;

app.use(express.json());
app.use(cors());

// Configuration
const ADMIN_API_URL = 'https://kd1s.com/admin/adminapi/v2/';
const ADMIN_API_KEY = '9495qsacgcm64hj0z1sprxce7qj0nkbh71w6h6xumigan8h141koirzkxwoqg48b';
const EXTERNAL_API_URL = 'https://kd1s.com/apikd1s';
const EXTERNAL_API_KEY = 'ce5d33dc71b144c60cab2f8f977bbc21';

const SERVICE_ID_TO_CHECK = 17337; // الخدمة المراد فحصها
const SERVICE_ID_TO_FORWARD = 17828; // الخدمة المراد إرسالها
const DUPLICATE_CHECK_WINDOW = 30 * 60 * 1000; // 30 دقيقة

// Data files
const PROCESSED_ORDERS_FILE = path.join(__dirname, 'processed_orders.json');
const USER_ORDERS_FILE = path.join(__dirname, 'user_orders.json');
const URL_ORDERS_FILE = path.join(__dirname, 'url_orders.json');

// Initialize files
function initializeFiles() {
    if (!fs.existsSync(PROCESSED_ORDERS_FILE)) {
        fs.writeFileSync(PROCESSED_ORDERS_FILE, JSON.stringify({}, null, 2));
    }
    if (!fs.existsSync(USER_ORDERS_FILE)) {
        fs.writeFileSync(USER_ORDERS_FILE, JSON.stringify({}, null, 2));
    }
    if (!fs.existsSync(URL_ORDERS_FILE)) {
        fs.writeFileSync(URL_ORDERS_FILE, JSON.stringify({}, null, 2));
    }
}

// Utility functions
function normalizeUrl(url) {
    try {
        // إزالة البروتوكول والـ www والمعاملات
        let normalized = url.replace(/^(https?:\/\/)?(www\.)?/, '').split('?')[0].split('#')[0];
        // إزالة الـ trailing slash
        normalized = normalized.replace(/\/$/, '');
        return normalized.toLowerCase();
    } catch (error) {
        return url.toLowerCase();
    }
}

function generateOrderHash(userId) {
    return crypto.createHash('md5').update(userId).digest('hex');
}

function loadJsonFile(filePath) {
    try {
        return JSON.parse(fs.readFileSync(filePath, 'utf8'));
    } catch (error) {
        return {};
    }
}

function saveJsonFile(filePath, data) {
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2));
}

// Check if user made an order within 30 minutes
function checkUserDuplicate(userId) {
    const userOrders = loadJsonFile(USER_ORDERS_FILE);
    const userHash = generateOrderHash(userId);
    
    if (userOrders[userHash]) {
        const lastOrderTime = userOrders[userHash];
        const timeSinceLastOrder = Date.now() - lastOrderTime;
        
        if (timeSinceLastOrder < DUPLICATE_CHECK_WINDOW) {
            return {
                isDuplicate: true,
                timeSinceLastOrder: timeSinceLastOrder,
                nextAllowedTime: lastOrderTime + DUPLICATE_CHECK_WINDOW
            };
        }
    }
    
    return { isDuplicate: false };
}

// Check if URL was used within 30 minutes
function checkUrlDuplicate(url) {
    const urlOrders = loadJsonFile(URL_ORDERS_FILE);
    const normalizedUrl = normalizeUrl(url);
    const urlHash = crypto.createHash('md5').update(normalizedUrl).digest('hex');
    
    if (urlOrders[urlHash]) {
        const lastOrderTime = urlOrders[urlHash];
        const timeSinceLastOrder = Date.now() - lastOrderTime;
        
        if (timeSinceLastOrder < DUPLICATE_CHECK_WINDOW) {
            return {
                isDuplicate: true,
                timeSinceLastOrder: timeSinceLastOrder,
                nextAllowedTime: lastOrderTime + DUPLICATE_CHECK_WINDOW
            };
        }
    }
    
    return { isDuplicate: false };
}

// Record user order
function recordUserOrder(userId) {
    const userOrders = loadJsonFile(USER_ORDERS_FILE);
    const userHash = generateOrderHash(userId);
    userOrders[userHash] = Date.now();
    saveJsonFile(USER_ORDERS_FILE, userOrders);
}

// Record URL order
function recordUrlOrder(url) {
    const urlOrders = loadJsonFile(URL_ORDERS_FILE);
    const normalizedUrl = normalizeUrl(url);
    const urlHash = crypto.createHash('md5').update(normalizedUrl).digest('hex');
    urlOrders[urlHash] = Date.now();
    saveJsonFile(URL_ORDERS_FILE, urlOrders);
}

// Fetch pending orders from admin API
async function fetchPendingOrders() {
    try {
        const response = await axios.get(`${ADMIN_API_URL}orders`, {
            params: {
                api_key: ADMIN_API_KEY,
                service: SERVICE_ID_TO_CHECK,
                status: 'Pending'
            },
            timeout: 30000
        });
        
        return response.data.data || [];
    } catch (error) {
        console.error('Error fetching pending orders:', error.message);
        return [];
    }
}

// Cancel and refund order
async function cancelAndRefundOrder(orderId) {
    try {
        const response = await axios.post(`${ADMIN_API_URL}orders/${orderId}/cancel`, {
            api_key: ADMIN_API_KEY
        }, {
            timeout: 30000
        });
        
        console.log(`Order ${orderId} cancelled and refunded`);
        return true;
    } catch (error) {
        console.error(`Error cancelling order ${orderId}:`, error.message);
        return false;
    }
}

// Forward order to external API
async function forwardOrderToExternalAPI(orderId, quantity, url) {
    try {
        const response = await axios.post(EXTERNAL_API_URL, {
            api_key: EXTERNAL_API_KEY,
            service: SERVICE_ID_TO_FORWARD,
            link: url,
            quantity: quantity
        }, {
            timeout: 30000
        });
        
        console.log(`Order ${orderId} forwarded to external API`);
        return {
            success: true,
            externalOrderId: response.data.order_id || orderId,
            response: response.data
        };
    } catch (error) {
        console.error(`Error forwarding order ${orderId}:`, error.message);
        return {
            success: false,
            error: error.message
        };
    }
}

// Process single order
async function processSingleOrder(order) {
    const orderId = order.id;
    const userId = order.user_id || order.username || 'unknown';
    const orderUrl = order.link || '';
    const quantity = order.quantity || 0;
    
    console.log(`\n=== Processing Order ${orderId} ===`);
    console.log(`User: ${userId}, URL: ${orderUrl}, Quantity: ${quantity}`);
    
    // Check if order was already processed
    const processedOrders = loadJsonFile(PROCESSED_ORDERS_FILE);
    if (processedOrders[orderId]) {
        console.log(`Order ${orderId} already processed`);
        return { status: 'skipped', reason: 'already_processed' };
    }
    
    // Step 1: Check user duplicate
    const userDuplicateCheck = checkUserDuplicate(userId);
    if (userDuplicateCheck.isDuplicate) {
        console.log(`User ${userId} has duplicate order within 30 minutes - CANCELLING`);
        await cancelAndRefundOrder(orderId);
        processedOrders[orderId] = {
            status: 'cancelled',
            reason: 'user_duplicate',
            timestamp: Date.now()
        };
        saveJsonFile(PROCESSED_ORDERS_FILE, processedOrders);
        return { status: 'cancelled', reason: 'user_duplicate' };
    }
    
    // Step 2: Check URL duplicate
    const urlDuplicateCheck = checkUrlDuplicate(orderUrl);
    if (urlDuplicateCheck.isDuplicate) {
        console.log(`URL ${orderUrl} has duplicate order within 30 minutes - CANCELLING`);
        await cancelAndRefundOrder(orderId);
        processedOrders[orderId] = {
            status: 'cancelled',
            reason: 'url_duplicate',
            timestamp: Date.now()
        };
        saveJsonFile(PROCESSED_ORDERS_FILE, processedOrders);
        return { status: 'cancelled', reason: 'url_duplicate' };
    }
    
    // Step 3: Forward to external API
    console.log(`Order ${orderId} passed all checks - forwarding to external API`);
    const forwardResult = await forwardOrderToExternalAPI(orderId, quantity, orderUrl);
    
    if (forwardResult.success) {
        // Record successful order
        recordUserOrder(userId);
        recordUrlOrder(orderUrl);
        
        processedOrders[orderId] = {
            status: 'forwarded',
            externalOrderId: forwardResult.externalOrderId,
            timestamp: Date.now()
        };
        saveJsonFile(PROCESSED_ORDERS_FILE, processedOrders);
        
        console.log(`Order ${orderId} successfully forwarded`);
        return { status: 'forwarded', externalOrderId: forwardResult.externalOrderId };
    } else {
        console.log(`Failed to forward order ${orderId}: ${forwardResult.error}`);
        return { status: 'failed', error: forwardResult.error };
    }
}

// Main processing function
async function processOrders() {
    console.log(`\n${'='.repeat(50)}`);
    console.log(`Processing started at ${new Date().toISOString()}`);
    console.log(`${'='.repeat(50)}`);
    
    try {
        const pendingOrders = await fetchPendingOrders();
        console.log(`Found ${pendingOrders.length} pending orders`);
        
        if (pendingOrders.length === 0) {
            console.log('No pending orders to process');
            return;
        }
        
        let successCount = 0;
        let cancelledCount = 0;
        let failedCount = 0;
        
        for (const order of pendingOrders) {
            const result = await processSingleOrder(order);
            
            if (result.status === 'forwarded') {
                successCount++;
            } else if (result.status === 'cancelled') {
                cancelledCount++;
            } else if (result.status === 'failed') {
                failedCount++;
            }
            
            // Add delay between requests to avoid rate limiting
            await new Promise(resolve => setTimeout(resolve, 1000));
        }
        
        console.log(`\n${'='.repeat(50)}`);
        console.log(`Processing completed`);
        console.log(`✅ Forwarded: ${successCount}`);
        console.log(`❌ Cancelled: ${cancelledCount}`);
        console.log(`⚠️ Failed: ${failedCount}`);
        console.log(`${'='.repeat(50)}\n`);
        
    } catch (error) {
        console.error('Error in processOrders:', error);
    }
}

// API Endpoints
app.get('/api/health', (req, res) => {
    res.json({
        status: 'ok',
        timestamp: new Date().toISOString(),
        service: 'order-processor',
        version: '1.0.0'
    });
});

app.get('/api/stats', (req, res) => {
    const processedOrders = loadJsonFile(PROCESSED_ORDERS_FILE);
    const userOrders = loadJsonFile(USER_ORDERS_FILE);
    const urlOrders = loadJsonFile(URL_ORDERS_FILE);
    
    const stats = {
        totalProcessed: Object.keys(processedOrders).length,
        totalUsers: Object.keys(userOrders).length,
        totalUrls: Object.keys(urlOrders).length,
        processedOrders: processedOrders,
        timestamp: new Date().toISOString()
    };
    
    res.json(stats);
});

app.post('/api/process-now', async (req, res) => {
    const key = req.body.key || req.query.key;
    
    if (key !== ADMIN_API_KEY) {
        return res.status(401).json({ error: 'Unauthorized' });
    }
    
    try {
        await processOrders();
        res.json({ status: 'processing_started' });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

app.get('/api/config', (req, res) => {
    res.json({
        SERVICE_ID_TO_CHECK,
        SERVICE_ID_TO_FORWARD,
        DUPLICATE_CHECK_WINDOW: `${DUPLICATE_CHECK_WINDOW / 60000} minutes`,
        ADMIN_API_URL,
        EXTERNAL_API_URL
    });
});

// Initialize and start server
initializeFiles();

app.listen(PORT, () => {
    console.log(`Order Processor Server running on port ${PORT}`);
    console.log(`Service ID to check: ${SERVICE_ID_TO_CHECK}`);
    console.log(`Service ID to forward: ${SERVICE_ID_TO_FORWARD}`);
    console.log(`Duplicate check window: ${DUPLICATE_CHECK_WINDOW / 60000} minutes`);
});

// Process orders every 5 minutes
setInterval(processOrders, 5 * 60 * 1000);

// Process orders on startup after 10 seconds
setTimeout(processOrders, 10 * 1000);

module.exports = app;

File: order_processor.py
#!/usr/bin/env python3
"""
Advanced Order Processing System with Duplicate Detection and Filtering
Language: Python 3
Framework: Flask
Database: PostgreSQL
"""

import os
import json
import hashlib
import re
from datetime import datetime, timedelta
from urllib.parse import urlparse
import requests
import psycopg2
from psycopg2.extras import RealDictCursor
from flask import Flask, request, jsonify
from flask_cors import CORS

# ============================================================================
# Configuration
# ============================================================================

app = Flask(__name__)
CORS(app)

# Database Configuration
DATABASE_URL = os.getenv('DATABASE_URL', 'postgresql://user:password@localhost/order_processor_db')

# API Configuration
ADMIN_API_URL = 'https://kd1s.com/admin/adminapi/v2/'
ADMIN_API_KEY = '9495qsacgcm64hj0z1sprxce7qj0nkbh71w6h6xumigan8h141koirzkxwoqg48b'

EXTERNAL_API_URL = 'https://kd1s.com/apikd1s'
EXTERNAL_API_KEY = 'ce5d33dc71b144c60cab2f8f977bbc21'

# Service Configuration
SOURCE_SERVICE_ID = 17337  # Service to fetch orders from
TARGET_SERVICE_ID = 17828  # Service to send orders to

# Security Configuration
USER_CHECK_WINDOW = 30 * 60  # 30 minutes in seconds
GLOBAL_COOLDOWN = 60 * 60   # 60 minutes in seconds

# ============================================================================
# Database Functions
# ============================================================================

def get_db_connection():
    """Create database connection"""
    try:
        conn = psycopg2.connect(DATABASE_URL)
        return conn
    except Exception as e:
        print(f"Database connection error: {e}")
        return None

def init_database():
    """Initialize database tables"""
    conn = get_db_connection()
    if not conn:
        return False
    
    try:
        cur = conn.cursor()
        
        # Create requests table
        cur.execute("""
            CREATE TABLE IF NOT EXISTS processed_requests (
                id SERIAL PRIMARY KEY,
                order_id VARCHAR(255) UNIQUE NOT NULL,
                user_hash VARCHAR(255) NOT NULL,
                url_hash VARCHAR(255) NOT NULL,
                original_url TEXT NOT NULL,
                service_id INTEGER NOT NULL,
                quantity INTEGER NOT NULL,
                status VARCHAR(50) DEFAULT 'pending',
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                processed_at TIMESTAMP,
                error_message TEXT,
                INDEX idx_user_hash (user_hash),
                INDEX idx_url_hash (url_hash),
                INDEX idx_created_at (created_at)
            )
        """)
        
        # Create statistics table
        cur.execute("""
            CREATE TABLE IF NOT EXISTS statistics (
                id SERIAL PRIMARY KEY,
                total_requests INTEGER DEFAULT 0,
                successful_requests INTEGER DEFAULT 0,
                failed_requests INTEGER DEFAULT 0,
                duplicate_requests INTEGER DEFAULT 0,
                last_update TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        """)
        
        conn.commit()
        cur.close()
        conn.close()
        print("✅ Database initialized successfully")
        return True
    except Exception as e:
        print(f"Database initialization error: {e}")
        return False

# ============================================================================
# Security Functions
# ============================================================================

def generate_hash(data):
    """Generate SHA256 hash"""
    return hashlib.sha256(data.encode()).hexdigest()

def extract_username_from_url(url):
    """Extract username from social media URL"""
    try:
        # Remove protocol and www
        url = re.sub(r'https?://(www\.)?', '', url)
        
        # Extract username based on platform
        if 'instagram.com' in url:
            match = re.search(r'instagram\.com/([^/?]+)', url)
            if match:
                return match.group(1)
        elif 'tiktok.com' in url:
            match = re.search(r'tiktok\.com/@([^/?]+)', url)
            if match:
                return match.group(1)
        elif 'telegram.org' in url or 't.me' in url:
            match = re.search(r'(?:t\.me|telegram\.org)/([^/?]+)', url)
            if match:
                return match.group(1)
        elif 'facebook.com' in url:
            match = re.search(r'facebook\.com/([^/?]+)', url)
            if match:
                return match.group(1)
        
        return url
    except:
        return url

def normalize_url(url):
    """Normalize URL to detect duplicates"""
    try:
        # Remove protocol
        url = re.sub(r'https?://', '', url)
        # Remove www
        url = re.sub(r'www\.', '', url)
        # Remove trailing slashes
        url = url.rstrip('/')
        # Convert to lowercase
        url = url.lower()
        # Remove query parameters
        url = url.split('?')[0]
        return url
    except:
        return url

def check_duplicate_user(user_hash, conn):
    """Check if user made request in last 30 minutes"""
    try:
        cur = conn.cursor(cursor_factory=RealDictCursor)
        
        cur.execute("""
            SELECT * FROM processed_requests 
            WHERE user_hash = %s 
            AND created_at > NOW() - INTERVAL '30 minutes'
            LIMIT 1
        """, (user_hash,))
        
        result = cur.fetchone()
        cur.close()
        
        return result is not None
    except Exception as e:
        print(f"Error checking duplicate user: {e}")
        return False

def check_duplicate_url(url_hash, conn):
    """Check if same URL was requested in last 30 minutes"""
    try:
        cur = conn.cursor(cursor_factory=RealDictCursor)
        
        cur.execute("""
            SELECT * FROM processed_requests 
            WHERE url_hash = %s 
            AND created_at > NOW() - INTERVAL '30 minutes'
            LIMIT 1
        """, (url_hash,))
        
        result = cur.fetchone()
        cur.close()
        
        return result is not None
    except Exception as e:
        print(f"Error checking duplicate URL: {e}")
        return False

# ============================================================================
# API Functions
# ============================================================================

def fetch_pending_orders():
    """Fetch pending orders from admin API"""
    try:
        headers = {'Authorization': f'Bearer {ADMIN_API_KEY}'}
        params = {
            'service': SOURCE_SERVICE_ID,
            'status': 'Pending'
        }
        
        response = requests.get(
            f"{ADMIN_API_URL}orders",
            headers=headers,
            params=params,
            timeout=10
        )
        
        if response.status_code == 200:
            return response.json().get('orders', [])
        else:
            print(f"Error fetching orders: {response.status_code}")
            return []
    except Exception as e:
        print(f"Error fetching orders: {e}")
        return []

def send_order_to_external_api(url, quantity):
    """Send validated order to external API"""
    try:
        headers = {'Authorization': f'Bearer {EXTERNAL_API_KEY}'}
        data = {
            'service': TARGET_SERVICE_ID,
            'link': url,
            'quantity': quantity
        }
        
        response = requests.post(
            EXTERNAL_API_URL,
            headers=headers,
            json=data,
            timeout=10
        )
        
        if response.status_code in [200, 201]:
            return True, response.json()
        else:
            return False, response.text
    except Exception as e:
        print(f"Error sending order: {e}")
        return False, str(e)

def cancel_order(order_id):
    """Cancel order and refund"""
    try:
        headers = {'Authorization': f'Bearer {ADMIN_API_KEY}'}
        data = {'action': 'cancel_and_refund'}
        
        response = requests.post(
            f"{ADMIN_API_URL}orders/{order_id}",
            headers=headers,
            json=data,
            timeout=10
        )
        
        return response.status_code in [200, 201]
    except Exception as e:
        print(f"Error canceling order: {e}")
        return False

# ============================================================================
# Processing Functions
# ============================================================================

def process_orders():
    """Main order processing function"""
    print(f"\n{'='*60}")
    print(f"🔄 Processing orders at {datetime.now()}")
    print(f"{'='*60}")
    
    conn = get_db_connection()
    if not conn:
        print("❌ Database connection failed")
        return
    
    try:
        # Fetch pending orders
        orders = fetch_pending_orders()
        print(f"📥 Fetched {len(orders)} pending orders")
        
        if not orders:
            print("✅ No pending orders")
            conn.close()
            return
        
        # Process each order
        for order in orders:
            try:
                order_id = order.get('id')
                url = order.get('link', '')
                quantity = order.get('quantity', 0)
                
                print(f"\n📋 Processing Order #{order_id}")
                print(f"   URL: {url}")
                print(f"   Quantity: {quantity}")
                
                # Generate hashes
                user_hash = generate_hash(order_id)
                normalized_url = normalize_url(url)
                url_hash = generate_hash(normalized_url)
                
                # Check for duplicate user
                if check_duplicate_user(user_hash, conn):
                    print(f"   ❌ DUPLICATE USER - Canceling and refunding")
                    cancel_order(order_id)
                    continue
                
                # Check for duplicate URL
                if check_duplicate_url(url_hash, conn):
                    print(f"   ❌ DUPLICATE URL - Canceling and refunding")
                    cancel_order(order_id)
                    continue
                
                # Send to external API
                print(f"   📤 Sending to external API...")
                success, response = send_order_to_external_api(url, quantity)
                
                if success:
                    print(f"   ✅ Order sent successfully")
                    
                    # Save to database
                    cur = conn.cursor()
                    cur.execute("""
                        INSERT INTO processed_requests 
                        (order_id, user_hash, url_hash, original_url, service_id, quantity, status)
                        VALUES (%s, %s, %s, %s, %s, %s, %s)
                    """, (order_id, user_hash, url_hash, url, TARGET_SERVICE_ID, quantity, 'completed'))
                    conn.commit()
                    cur.close()
                else:
                    print(f"   ❌ Failed to send order: {response}")
                    
                    # Save error to database
                    cur = conn.cursor()
                    cur.execute("""
                        INSERT INTO processed_requests 
                        (order_id, user_hash, url_hash, original_url, service_id, quantity, status, error_message)
                        VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
                    """, (order_id, user_hash, url_hash, url, SOURCE_SERVICE_ID, quantity, 'failed', str(response)))
                    conn.commit()
                    cur.close()
            
            except Exception as e:
                print(f"   ❌ Error processing order: {e}")
                continue
        
        print(f"\n{'='*60}")
        print(f"✅ Order processing completed")
        print(f"{'='*60}\n")
    
    except Exception as e:
        print(f"❌ Error in process_orders: {e}")
    
    finally:
        conn.close()

# ============================================================================
# Flask Routes
# ============================================================================

@app.route('/api/health', methods=['GET'])
def health():
    """Health check endpoint"""
    conn = get_db_connection()
    db_status = "connected" if conn else "disconnected"
    if conn:
        conn.close()
    
    return jsonify({
        'status': 'ok',
        'timestamp': datetime.now().isoformat(),
        'database': db_status,
        'service': 'order-processor'
    })

@app.route('/api/stats', methods=['GET'])
def stats():
    """Get statistics"""
    conn = get_db_connection()
    if not conn:
        return jsonify({'error': 'Database connection failed'}), 500
    
    try:
        cur = conn.cursor(cursor_factory=RealDictCursor)
        
        cur.execute("""
            SELECT 
                COUNT(*) as total,
                SUM(CASE WHEN status = 'completed' THEN 1 ELSE 0 END) as successful,
                SUM(CASE WHEN status = 'failed' THEN 1 ELSE 0 END) as failed
            FROM processed_requests
        """)
        
        stats = cur.fetchone()
        cur.close()
        conn.close()
        
        return jsonify({
            'total_requests': stats['total'] or 0,
            'successful_requests': stats['successful'] or 0,
            'failed_requests': stats['failed'] or 0,
            'timestamp': datetime.now().isoformat()
        })
    
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@app.route('/api/config', methods=['GET'])
def config():
    """Get configuration"""
    return jsonify({
        'source_service': SOURCE_SERVICE_ID,
        'target_service': TARGET_SERVICE_ID,
        'user_check_window': USER_CHECK_WINDOW,
        'global_cooldown': GLOBAL_COOLDOWN,
        'timestamp': datetime.now().isoformat()
    })

@app.route('/', methods=['GET'])
def index():
    """Index endpoint"""
    return jsonify({
        'name': 'Order Processor',
        'version': '1.0.0',
        'language': 'Python',
        'framework': 'Flask',
        'database': 'PostgreSQL',
        'endpoints': {
            'health': '/api/health',
            'stats': '/api/stats',
            'config': '/api/config'
        }
    })

# ============================================================================
# Main
# ============================================================================

if __name__ == '__main__':
    print("🚀 Starting Order Processor (Python/Flask)")
    
    # Initialize database
    init_database()
    
    # Start Flask app
    port = int(os.getenv('PORT', 10001))
    app.run(host='0.0.0.0', port=port, debug=False)
